network/validator/pubsub/authorized_sender_validator_test.go

```diff
@@ -0,0 +1,120 @@
+	suite.Suite
+	happyPathTestCases []TestCase
+	sadPathTestCases   []TestCase
+}
+
+func (s *TestIsAuthorizedSenderSuite) SetupTest() {
+	s.initializeTestCases()
+}
+
+// TestIsAuthorizedSender_AuthorizedSender checks that IsAuthorizedSender does not return false positive
+// validation errors for all possible valid combinations (authorized sender role, message type).
+func (s *TestIsAuthorizedSenderSuite) TestIsAuthorizedSender_AuthorizedSender() {
+	for _, c := range s.happyPathTestCases {
+		str := fmt.Sprintf("role (%s) should be authorized to send message type (%s) on channel (%s)", c.Identity.Role, c.MessageStr, c.Channel)
+		s.Run(str, func() {
+			msgType, err := IsAuthorizedSender(c.Identity, c.Channel, c.Message)
+			s.Require().NoError(err)
+			s.Require().Equal(c.MessageStr, msgType)
+		})
+	}
+}
+
+// TestIsAuthorizedSender_UnAuthorizedSender checks that IsAuthorizedSender return's ErrUnauthorizedSender
+// validation error for all possible invalid combinations (unauthorized sender role, message type).
+func (s *TestIsAuthorizedSenderSuite) TestIsAuthorizedSender_UnAuthorizedSender() {
+	for _, c := range s.sadPathTestCases {
+		str := fmt.Sprintf("role (%s) should not be authorized to send message type (%s) on channel (%s)", c.Identity.Role, c.MessageStr, c.Channel)
+		s.Run(str, func() {
+			msgType, err := IsAuthorizedSender(c.Identity, c.Channel, c.Message)
+			s.Require().ErrorIs(err, ErrUnauthorizedSender)
+			s.Require().Equal(c.MessageStr, msgType)
+		})
+	}
+}
+
+// TestIsAuthorizedSender_ValidationFailure checks that IsAuthorizedSender returns the expected validation error.
+func (s *TestIsAuthorizedSenderSuite) TestIsAuthorizedSender_ValidationFailure() {
+	s.Run("sender is ejected", func() {
+		identity := unittest.IdentityFixture()
+		identity.Ejected = true
+		msgType, err := IsAuthorizedSender(identity, network.Channel(""), nil)
+		s.Require().ErrorIs(err, ErrSenderEjected)
+		s.Require().Equal("", msgType)
+	})
+
+	s.Run("unknown message type", func() {
+		identity := unittest.IdentityFixture()
+		msgType, err := IsAuthorizedSender(identity, network.Channel(""), nil)
+		s.Require().ErrorIs(err, ErrUnknownMessageType)
+		s.Require().Equal("", msgType)
+	})
+
+	s.Run("unknown message type with message embedded", func() {
+		identity := unittest.IdentityFixture(unittest.WithRole(flow.RoleConsensus))
+		type msg struct {
+			*messages.BlockProposal
+		}
+
+		m := &msg{&messages.BlockProposal{
+			Header:  nil,
+			Payload: nil,
+		}}
+
+		msgType, err := IsAuthorizedSender(identity, network.ConsensusCommittee, m)
+		s.Require().ErrorIs(err, ErrUnknownMessageType)
+		s.Require().Equal("", msgType)
+	})
+}
+
+// initializeTestCases initializes happy and sad path test cases for checking authorized and unauthorized role message combinations.
+func (s *TestIsAuthorizedSenderSuite) initializeTestCases() {
+	for _, c := range network.MessageAuthConfigs {
+		for channel, authorizedRoles := range c.Config {
```

Comment:
Currently, unhappy path cases concern the wrong sender role (but not message type), right?. Can we also iterate on all message authentication configs and hence have unhappy path test cases including message types as well? i.e., sending messages on the wrong channels.

Reply:
https://github.com/onflow/flow-go/pull/2648/commits/f848585d4d1623be7953c851c92ad7fb53e19909